compiler/src/dotty/tools/dotc/core/Types.scala

```diff
@@ -598,11 +588,20 @@ object Types {
         case AndType(l, r) =>
           goAnd(l, r)
         case tp: OrType =>
-          // we need to keep the invariant that `pre <: tp`. Branch `union-types-narrow-prefix`
-          // achieved that by narrowing `pre` to each alternative, but it led to merge errors in
-          // lots of places. The present strategy is instead of widen `tp` using `join` to be a
-          // supertype of `pre`.
-          go(tp.join)
+          if (ctx.explicitNulls && tp.isJavaNullableUnion) {
+            // Selecting `name` from a type `T|JavaNull` is like selecting `name` from `T`.
```

Comment:
It seems isJavaNullable union and stripJavaNull do the same core operation twice. It might be more efficient to use a `OrJavaNull` extractor instead (which should succeed only under ctx.explicitNulls):
```
  tp match
    case OrJavaNull(tp1) => tp1.findMember(name, pre.stripJavaNull, required, excluded)
    case _ => go(tp.join)
```

Reply:
`OrJavaNull` is added.